### PR TITLE
Bootstrapper plugin support and Kubernetes bootstrapper

### PR DESCRIPTION
**Pull request by azdagron**
Implements support for bootstrapper plugin type as well as the bootstrapping flow in the server CA manager.

A Kubernetes bootstrapper plugin is also included that publishes bundle updates to a ConfigMap.

See the comments in proto/server/bootstrapper/bootstrapper.proto for details on the flow.

Also took this opportunity to swap out the "now" hook in the CA manager with a Clock to ease testing.

### CODE REVIEW

File: pkg/server/plugin/bootstrapper/k8sbootstrapper/k8sbootstrapper.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	bootstrapErr = errs.Class("k8s-bootstrapper")
+)
+
+const (
+	defaultNamespace    = "spire"
+	defaultConfigMap    = "spire-bootstrap"
+	defaultConfigMapKey = "bootstrap.crt"
+)
+
+type ConfigUpdaterConfig struct {
```
Comment (evan2645) [inline pkg/server/plugin/bootstrapper/k8sbootstrapper/k8sbootstrapper.go]: Does this need to be exported?

File: pkg/server/plugin/bootstrapper/k8sbootstrapper/k8sbootstrapper.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+		for _, rootCA := range bundle.RootCas {
+			pem.Encode(bundleData, &pem.Block{
+				Type:  "CERTIFICATE",
+				Bytes: rootCA.DerBytes,
+			})
+		}
+
+		// Replace the config map data
+		configMap.Data = map[string]string{
+			config.ConfigMapKey: bundleData.String(),
```
Comment (evan2645) [inline pkg/server/plugin/bootstrapper/k8sbootstrapper/k8sbootstrapper.go]: Would this overwrite any existing data in the config map aside from the `bootstrap.crt` key?

Comment (azdagron) [reply pkg/server/plugin/bootstrapper/k8sbootstrapper/k8sbootstrapper.go]: I implemented patch first but then moved to update when some documentation lead me to believe the ResourceVersion-based conflict resolution didn't work with patch. I've found some more commentary suggesting it does work. I do think patch is the right operation, if it works with our conflict resolution strategy. I'll move back to patch.

File: pkg/server/plugin/bootstrapper/k8sbootstrapper/k8sbootstrapper.go (hunk 1 above)
Comment (azdagron) [reply pkg/server/plugin/bootstrapper/k8sbootstrapper/k8sbootstrapper.go]: Nope. Other plugins also have exported config for no good reason :)

File: pkg/server/catalog/catalog.go [hunk 3]
```diff
@@ -248,6 +269,9 @@ func (c *ServerCatalog) categorize() error {
 			return fmt.Errorf("At least one plugin of type %s is required", t)
 		}
 	}
+	if len(c.bootstrapperPlugins) > 1 {
+		return fmt.Errorf("Only one Bootstrapper plugin can be enabled.")
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/server/catalog/catalog.go]: Hmm. Sure, I can see that. I'll lift the restriction and update the manager code.

File: pkg/server/catalog/catalog.go [hunk 4]
```diff
@@ -82,6 +87,9 @@ var (
 			"disk":   keymanager.NewBuiltIn(keymanager_disk.New()),
 			"memory": keymanager.NewBuiltIn(keymanager_memory.New()),
 		},
+		BootstrapperType: {
+			"k8s_bootstrapper": bootstrapper.NewBuiltIn(k8sbootstrapper.New()),
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/server/catalog/catalog.go]: A good nit.

File: pkg/server/ca/manager.go [hunk 5]
```diff
@@ -417,6 +447,72 @@ func (m *manager) prepareKeypairSet(ctx context.Context, kps *keypairSet) (err e
...
+	defer stream.CloseSend()
+
+	// the bootstrapper will send back a response as soon as it is ready
+	// to publish.
+	if _, err = stream.Recv(); err != nil {
+		return err
+	}
+
+	for {
+		m.c.Log.Debug("Bootstrapper is ready to publish bundle.")
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/server/ca/manager.go]: A good nit.

**Comment by evan2645** [comment]
@azdagron - in light of ongoing work to update the catalog and introduce `Notifier` plugin that will solve for this use case, should we close this PR out?

**Comment by azdagron** [comment]
Yup. That sounds good.

**Pull request closed without merging**